### PR TITLE
Visual song action

### PR DESCRIPTION
**Pull request by Octoton**
I am not a fan of the visual used for the new song click action when queue is not empty. This is a proposal of another way to show it to the user (see below).

Moreover, I have done some correction:
 - playing without queue or play action with a queue will both do the same => replace queue with all song of the playlist and play at the chosen position
 - make the new toast viewable on android 13 (instead of white text on white background in dark mode)
 - static queue implementation will now manage shuffle action all the time (adding songs, opening queue, ...)

Screenshot_20230205-165642

### CODE REVIEW

**Comment by Octoton** [comment]
@soncaokim, @AdrienPoupa could you look at this ?

**Comment by soncaokim** [comment]
I like the visual. Let me play with this a little bit.

**Comment by soncaokim** [comment]
In mid/long term, I guess this would make sense to adopt this bottom sheet and phasing out other dialogs?

**Comment by soncaokim** [comment]
Would it make sense to animate a countdown of N seconds on the default/last chosen option?

After the count down, the dialog is dismissed and the default action executed

**Comment by Octoton** [comment]
> Would it make sense to animate a countdown of N seconds on the default/last chosen option?
> 
> After the count down, the dialog is dismissed and the default action executed

I am really ensure about this, I don't think it will help people that want this to go fast (as they just need to click on a ligne) and it could stress people that like to go slow with UI.

**Comment by Octoton** [comment]
> In mid/long term, I guess this would make sense to adopt this bottom sheet and phasing out other dialogs?

Yes I imagine, I wanted to use this for settings of random album at a later date too.
However, not really sure what is the limit, is it all the dialog ? All the sub-menu ?

**Comment by soncaokim** [comment]
> I am really unsure about this, I don't think it will help people that want this to go fast (as they just need to click on a ligne) and it could stress people that like to go slow with UI.

In this case, there is no need to save/restore the default choice in the Prefs.

**Comment by soncaokim** [comment]
Can you please apply the same color of Snackbar button (the Undo button) to the 3 buttons in the bottom sheet?

I.e visual indication that these are actionable choices

**Comment by Octoton** [comment]
> > I am really unsure about this, I don't think it will help people that want this to go fast (as they just need to click on a ligne) and it could stress people that like to go slow with UI.
> 
> In this case, there is no need to save/restore the default choice in the Prefs.

Yes I was keeping the code just in case we wanted to add a option to no always show the menu (as it is not that big of dead code)

File: app/src/main/java/com/poupa/vinylmusicplayer/dialogs/BottomSheetDialog/BottomSheetDialog.java
Comment (soncaokim) [inline app/src/main/java/com/poupa/vinylmusicplayer/dialogs/BottomSheetDialog/BottomSheetDialog.java:40]: You cannot avoid having an empty overriden function?

File: app/src/main/java/com/poupa/vinylmusicplayer/dialogs/BottomSheetDialog/ButtonListBottomSheetDialog.java
Comment (soncaokim) [inline app/src/main/java/com/poupa/vinylmusicplayer/dialogs/BottomSheetDialog/ButtonListBottomSheetDialog.java:22]: Would 'BottomSheetDialogWithButtons' be a better name?

File: app/src/main/java/com/poupa/vinylmusicplayer/dialogs/BottomSheetDialog/EnqueueSongsBottomSheetDialog.java
Comment (soncaokim) [inline app/src/main/java/com/poupa/vinylmusicplayer/dialogs/BottomSheetDialog/EnqueueSongsBottomSheetDialog.java:9]: Suggestion: Name can be shortened to 'EnqueueSongsDialog'

File: app/src/main/java/com/poupa/vinylmusicplayer/dialogs/BottomSheetDialog/EnqueueSongsBottomSheetDialog.java
Comment (soncaokim) [inline app/src/main/java/com/poupa/vinylmusicplayer/dialogs/BottomSheetDialog/EnqueueSongsBottomSheetDialog.java:20]: I dont get the reason of this override: you just call super's implementation, in this case there is no need for override, exact?

File: app/src/main/java/com/poupa/vinylmusicplayer/dialogs/BottomSheetDialog/EnqueueSongsBottomSheetDialog.java
Comment (soncaokim) [inline app/src/main/java/com/poupa/vinylmusicplayer/dialogs/BottomSheetDialog/EnqueueSongsBottomSheetDialog.java:26]: Same here (useless override?)

File: app/src/main/java/com/poupa/vinylmusicplayer/dialogs/BottomSheetDialog/EnqueueSongsBottomSheetDialog.java
Comment (soncaokim) [inline app/src/main/java/com/poupa/vinylmusicplayer/dialogs/BottomSheetDialog/EnqueueSongsBottomSheetDialog.java:33]: As discussed, we dont need to save and restore the default choice

File: app/src/main/java/com/poupa/vinylmusicplayer/misc/queue/StaticPlayingQueue.java
Comment (soncaokim) [inline app/src/main/java/com/poupa/vinylmusicplayer/misc/queue/StaticPlayingQueue.java:120]: I'd prefer having braces {} here

File: app/src/main/java/com/poupa/vinylmusicplayer/misc/queue/StaticPlayingQueue.java
Comment (soncaokim) [inline app/src/main/java/com/poupa/vinylmusicplayer/misc/queue/StaticPlayingQueue.java:208]: Braces {} please

File: app/src/main/java/com/poupa/vinylmusicplayer/dialogs/BottomSheetDialog/BottomSheetDialog.java
Comment (Octoton) [reply app/src/main/java/com/poupa/vinylmusicplayer/dialogs/BottomSheetDialog/BottomSheetDialog.java:40]: BottomSheetBehavior.BottomSheetCallback is abstract and need a slide declaration

File: app/src/main/java/com/poupa/vinylmusicplayer/dialogs/BottomSheetDialog/EnqueueSongsBottomSheetDialog.java
Comment (Octoton) [reply app/src/main/java/com/poupa/vinylmusicplayer/dialogs/BottomSheetDialog/EnqueueSongsBottomSheetDialog.java:20]: I need this or setTitle will return a BottomSheetDialogWithButtons object and I will not be able to call setDefaultChoice after calling setTitle
Not sure if the override is necessary, seems cleaner at least

File: app/src/main/java/com/poupa/vinylmusicplayer/dialogs/BottomSheetDialog/EnqueueSongsBottomSheetDialog.java
Comment (Octoton) [reply app/src/main/java/com/poupa/vinylmusicplayer/dialogs/BottomSheetDialog/EnqueueSongsBottomSheetDialog.java:26]: same thing

File: app/src/main/java/com/poupa/vinylmusicplayer/dialogs/BottomSheetDialog/EnqueueSongsBottomSheetDialog.java
Comment (Octoton) [reply app/src/main/java/com/poupa/vinylmusicplayer/dialogs/BottomSheetDialog/EnqueueSongsBottomSheetDialog.java:33]: > Yes I was keeping the code just in case we wanted to add a option to no always show the menu (as it is not that big of dead code)

So we don't want to keep this just in case ?

**Comment by Octoton** [comment]
> Can you please apply the same color of Snackbar button (the Undo button) to the 3 buttons in the bottom sheet?
> 
> I.e visual indication that these are actionable choices

Wasn't able to make this work for know, will try again later in the week
As a side note, the existing implementation seems quite strange (the undo color), seems to only work when theme is light and I cannot find out why

File: app/src/main/java/com/poupa/vinylmusicplayer/dialogs/BottomSheetDialog/EnqueueSongsBottomSheetDialog.java
Comment (soncaokim) [reply app/src/main/java/com/poupa/vinylmusicplayer/dialogs/BottomSheetDialog/EnqueueSongsBottomSheetDialog.java]: I'd be aggressive and drop it now, before a release is made (otherwise it will become legacy)

**Comment by soncaokim** [comment]
> Wasn't able to make this work for know, will try again later in the week

On second though, the coloring suggestion I've made doesnt seem to be consistent with menu and other buttons.
Maybe just increase the font size a bit if you dont mind

File: app/src/main/java/com/poupa/vinylmusicplayer/dialogs/BottomSheetDialog/BottomSheetDialogWithButtons.java
Comment (soncaokim) [inline app/src/main/java/com/poupa/vinylmusicplayer/dialogs/BottomSheetDialog/BottomSheetDialogWithButtons.java:94]: Can you please rename this to Button or ButtonInfo, i.e matching the class name BottomSheetWithButtons

File: app/src/main/java/com/poupa/vinylmusicplayer/dialogs/BottomSheetDialog/EnqueueSongsBottomSheetDialog.java
Comment (soncaokim) [reply app/src/main/java/com/poupa/vinylmusicplayer/dialogs/BottomSheetDialog/EnqueueSongsBottomSheetDialog.java]: This still bothers me, but I can live with it for now.
Probably when we want to apply this bottom sheet to other dialogs, we need to revisit this to avoid (possible) proliferation of trivial overriden methods.

**Comment by soncaokim** [comment]
What do you think about a bottom sheet like the one below (from https://m3.material.io/components/bottom-sheets/guidelines), where:
- there is a small divider separating the message and possible actions
- the actions have icons 

image

**Review by soncaokim — APPROVED** [review]

**Comment by Octoton** [comment]
> What do you think about a bottom sheet like the one below (from https://m3.material.io/components/bottom-sheets/guidelines), where:
> 
>     * there is a small divider separating the message and possible actions
> 
>     * the actions have icons
> 
> 
> image

This look really slick, I was also thinking about adding icon
I will look into this

**Comment by soncaokim** [comment]
@Octoton Do you want to merge as it is, and do a separate PR for the bottom sheet with icons;
or you prefer having them all together?

I'd suggest you to merge right now just to avoid potential conflict.

**Comment by Octoton** [comment]
Sorry about the delay, didn't had much time after work to look at this.
After, trying some style out, I have landed on a simple one (see below), and without a divider line as it doesn't work quite well when there is only a text before it (didn't see the point of adding a image on the pop-up as we are managing a list of song and not a unique one - perhaps when we migrate the other dialog to this style we can change this).

Screenshot_20230211-151608

**Comment by soncaokim** [comment]
Top moumoute 😁
Let me have a try with it, thanks

**Comment by soncaokim** [comment]
Good for me, thanks!

**Comment by soncaokim** [comment]
Some improvement ideas:
- With the Black (amoled) theme, is it possible to have a slight grey border so that user can see the bottom sheet separated from the other elements on screen?
- Would it make sense to make the bottom sheet:s corner rounded if the rounded theme is active, and straight corner if classic theme is active?

**Comment by soncaokim** [comment]
And can you please increase the font size, ie use the same as the menu item/dialog default text?

**Comment by soncaokim** [comment]
Actually, there are already existing icons associated with Play next/Add to queue according to

https://github.com/VinylMusicPlayer/VinylMusicPlayer/blob/3a8f8074ac4df5df6341aea4b8d8e89797495ce4/app/src/main/res/menu/menu_cannot_delete_single_songs_playlist_songs_selection.xml#L7

and

https://github.com/VinylMusicPlayer/VinylMusicPlayer/blob/3a8f8074ac4df5df6341aea4b8d8e89797495ce4/app/src/main/res/menu/menu_cannot_delete_single_songs_playlist_songs_selection.xml#L13

Same in `menu_media_selection.xml`, `menu_playlists_selection.xml`, `menu_playlists_songs_selection.xml`

-> can you please use the same icons for the menu and for the bottom sheet (I think your choice of icon is better though)

**Review by soncaokim — CHANGES_REQUESTED** [review]
Actually, there are already existing icons associated with Play next/Add to queue according to

https://github.com/VinylMusicPlayer/VinylMusicPlayer/blob/3a8f8074ac4df5df6341aea4b8d8e89797495ce4/app/src/main/res/menu/menu_cannot_delete_single_songs_playlist_songs_selection.xml#L7

and

https://github.com/VinylMusicPlayer/VinylMusicPlayer/blob/3a8f8074ac4df5df6341aea4b8d8e89797495ce4/app/src/main/res/menu/menu_cannot_delete_single_songs_playlist_songs_selection.xml#L13

Same in `menu_media_selection.xml`, `menu_playlists_selection.xml`, `menu_playlists_songs_selection.xml`

-> can you please use the same icons for the menu and for the bottom sheet (I think your choice of icon is better though)

**Comment by Octoton** [comment]
Should answer to all your suggestion (I have gone back to the existing icon as to not change user habit)
Moreover, I have use your removeDuplicate also on menu use to synchronized the user experiences

The screenshoot show a separator that is almost not visible, however on my phone it is better (don't know how much each screen can change this)
Screenshot_20230212-230653

**Review by soncaokim — APPROVED** [review]
Looks good to me

**Comment by soncaokim** [comment]
Can we merge this in? 
I'd like to prep a new 1.5.0 release, we've accumulated a bunch of changes lately.

**Pull request merged**
